### ISSUE TITLE
Dynamic RedirectUris

### ISSUE THREAD

**Issue by johnkors** [cobraz/nextjs-oidc-react#2]
Nice! I'm also looking into using oidc-js in Next.js. 

One thing I stumbled upon when deploying to Zeit using Now, is the need for the redirectUri to be dynamically set - since the preview sites in Now are pretty dynamic. Do you have any idea on how to implement that?

I've so far fiddled with making the oidc redirecturi  be dynamic based on the host header of the incoming request.

https://github.com/cobraz/nextjs-oidc-react/blob/master/components/Auth.tsx#L18

**Comment by cobraz** [comment]
Hello 👋

Thanks for reaching out!

Have you tried using `process.env.NOW_URL`?

```typescript
const oidcConfig = {
  (...)
  redirectUri:
    process.env.NODE_ENV === "development"
      ? "http://localhost:3000/"
      : process.env.NOW_URL,
};
```

I haven't tested it. Maybe you could try it out, and let me know how it works?
I'm thinking about making a tutorial on Zeit Now (and others) eventually. 😃

**Comment by johnkors** [comment]
Yeah, I tried that first. It's not present for some reason, even if enabling in the environment in the Now project — so not entirely sure why it's not working. Some forums mention reading it from a custom header instead.

Mark, I'm not a React/Next.js nor js developer by trade, so my approach might be stupid. I'm currently trying to do this in a custom `_app.js` setup and use the `UserManager` from `oidc-js` in other component/pages.

```js
static getInitialProps = function ({ ctx }) {

    let baseUrl;
    if (ctx.req) {
        // Server side rendering
        let protocol = "http";
        if (ctx.req.headers.referer)
            protocol = ctx.req.headers.referer.split('://')[0];
        baseUrl = `${protocol}://${ctx.req.headers.host}`
    } else {    
        baseUrl = `${window.location.protocol}//${window.location.hostname}`
    }

    const oidcSettings = createOidcSettings(baseUrl); 
    const manager = new UserManager(oidcSettings);
    manager.events.addUserLoaded(this.onUserLoaded.bind(this));

    return { 
        manager: manager
    };
}
```

I probably should look into using auth via hooks, but trying to learn step-by-step.

Look forward to a Now tutorial, that sounds great!

**Comment by cobraz** [comment]
Hmm. Yeah, I can see there are multiple issues about `NOW_URL` is missing. With this you are getting a dynamic URL or not?

You know what, I'll have a look at this in the next couple of days. I suggest you check out `oidc-react` btw, it should make it easier to use `oidc-client` 👍

**Comment by johnkors** [comment]
Yup, seem to work to get the URL. 

Was actually hoping something like `oidc-react` was available - the only reason I'm using `oidc-client` was because I could not find any thing like that some weeks ago. Great timing! Thanks for creating it!

**Comment by johnkors** [comment]
Struggling to get it working following this sample repo here. Would you expect the following to work..? I'm seeing `window is not defined` errors when I do this. I see you have something like `ssr: false` in the samples, but not sure how to do this without it. 

```js
const Auth = props => {
    return <AuthProvider {...props.oidcSettings}>{props.children}</AuthProvider>;
};

Auth.getInitialProps = async (ctx) => {
    console.log(inspect(ctx));

    let baseUrl;
    if (ctx.req) {
        // Server side rendering
        let protocol = "http";
        if (ctx.req.headers.referer)
            protocol = ctx.req.headers.referer.split('://')[0];

        baseUrl = `${protocol}://${ctx.req.headers.host}`
    } else if (window) {
        baseUrl = `${window.location.protocol}//${window.location.hostname}`
    }
    var oidcSettings = getOidcConfig(baseUrl);
    return {
        oidcSettings: oidcSettings
    }
}
```
